File: macosapp/localauth/localauth.go
// Package localauth resolves Bearer tokens for the local macOS menu-bar app.
// Resolution order: local-agent-config.json → server-credentials → none.
// Fall-through only on missing/empty/invalid sources (not HTTP 401).
package localauth

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
)

// TokenSource identifies where ResolveLocalServerToken obtained the token.
type TokenSource string

const (
	SourceConfig      TokenSource = "config"
	SourceCredentials TokenSource = "credentials"
	SourceNone        TokenSource = "none"
)

// Options controls where resolve looks for files.
// DataDir empty defaults to ~/.ai-critic.
// ConfigPath / CredentialsPath override the default file paths under DataDir when set.
type Options struct {
	DataDir         string
	ConfigPath      string
	CredentialsPath string
}

// configFile is the on-disk local-agent CLI schema.
type configFile struct {
	Default string         `json:"default"`
	Domains []configDomain `json:"domains"`
}

type configDomain struct {
	Server string `json:"server"`
	Token  string `json:"token"`
}

// Local loopback servers the menu-bar app targets (after normalize).
var localLoopbackServers = []string{
	"http://localhost:23712",
	"http://127.0.0.1:23712",
}

// ResolveLocalServerToken returns a bearer token and its source for the local
// loopback server. Missing files, invalid JSON, and empty tokens fall through;
// resolve never returns an error.
func ResolveLocalServerToken(opts Options) (token string, source TokenSource) {
	dataDir := opts.DataDir
	if dataDir == "" {
		home, err := os.UserHomeDir()
		if err != nil || home == "" {
			// Cannot resolve home; still try credentials path if absolute overrides set.
			dataDir = ""
		} else {
			dataDir = filepath.Join(home, ".ai-critic")
		}
	}

	configPath := opts.ConfigPath
	if configPath == "" {
		if dataDir != "" {
			configPath = filepath.Join(dataDir, "local-agent-config.json")
		}
	}
	credsPath := opts.CredentialsPath
	if credsPath == "" {
		if dataDir != "" {
			credsPath = filepath.Join(dataDir, "server-credentials")
		}
	}

	if configPath != "" {
		if t, ok := tokenFromConfig(configPath); ok {
			return t, SourceConfig
		}
	}
	if credsPath != "" {
		if t, ok := tokenFromCredentials(credsPath); ok {
			return t, SourceCredentials
		}
	}
	return "", SourceNone
}

// AuthorizationHeader formats a Bearer token header value.
// Empty token returns "" (caller should omit the header).
func AuthorizationHeader(token string) string {
	if token == "" {
		return ""
	}
	return "Bearer " + token
}

// NormalizeServer trims space and trailing slashes (same contract as remoteconfig).
func NormalizeServer(server string) string {
	return strings.TrimRight(strings.TrimSpace(server), "/")
}

func tokenFromConfig(path string) (string, bool) {
	data, err := os.ReadFile(path)
	if err != nil {
		return "", false
	}
	var cfg configFile
	if err := json.Unmarshal(data, &cfg); err != nil {
		return "", false
	}

	// 1) Prefer a local loopback domain with a non-empty trimmed token.
	for _, target := range localLoopbackServers {
		if t, ok := domainTokenMatching(cfg.Domains, target); ok {
			return t, true
		}
	}

	// 2) Else use the domain matching default (after normalize).
	def := NormalizeServer(cfg.Default)
	if def != "" {
		if t, ok := domainTokenMatching(cfg.Domains, def); ok {
			return t, true
		}
	}
	return "", false
}

func domainTokenMatching(domains []configDomain, wantNormalized string) (string, bool) {
	for _, d := range domains {
		if NormalizeServer(d.Server) != wantNormalized {
			continue
		}
		t := strings.TrimSpace(d.Token)
		if t != "" {
			return t, true
		}
	}
	return "", false
}

func tokenFromCredentials(path string) (string, bool) {
	data, err := os.ReadFile(path)
	if err != nil {
		return "", false
	}
	for _, line := range strings.Split(string(data), "\n") {
		t := strings.TrimSpace(line)
		if t != "" {
			return t, true
		}
	}
	return "", false
}

File: macos-ai-critic/ai-critic-macos/LocalAuth.swift
import Foundation

/// Pure helpers for resolving a Bearer token for the local loopback server.
/// Mirrors `macosapp/localauth`: config first, then server-credentials first line.
enum LocalAuth {
    static let configFileName = "local-agent-config.json"
    static let credentialsFileName = "server-credentials"

    /// Local loopback servers after normalize (trim space + trailing `/`).
    static let localLoopbackServers = [
        "http://localhost:23712",
        "http://127.0.0.1:23712",
    ]

    enum TokenSource: String {
        case config
        case credentials
        case none
    }

    struct Options {
        var dataDir: String = ""
        var configPath: String = ""
        var credentialsPath: String = ""
    }

    private struct ConfigFile: Decodable {
        var defaultServer: String
        var domains: [Domain]

        enum CodingKeys: String, CodingKey {
            case defaultServer = "default"
            case domains
        }

        init(from decoder: Decoder) throws {
            let c = try decoder.container(keyedBy: CodingKeys.self)
            defaultServer = try c.decodeIfPresent(String.self, forKey: .defaultServer) ?? ""
            domains = try c.decodeIfPresent([Domain].self, forKey: .domains) ?? []
        }
    }

    private struct Domain: Decodable {
        var server: String
        var token: String

        enum CodingKeys: String, CodingKey {
            case server, token
        }

        init(from decoder: Decoder) throws {
            let c = try decoder.container(keyedBy: CodingKeys.self)
            server = try c.decodeIfPresent(String.self, forKey: .server) ?? ""
            token = try c.decodeIfPresent(String.self, forKey: .token) ?? ""
        }
    }

    /// Resolve token from `~/.ai-critic` (or opts.dataDir): config → credentials → none.
    static func resolveLocalServerToken(opts: Options = Options()) -> (token: String, source: TokenSource) {
        var dataDir = opts.dataDir
        if dataDir.isEmpty {
            dataDir = (NSHomeDirectory() as NSString).appendingPathComponent(".ai-critic")
        }

        let configPath = opts.configPath.isEmpty
            ? (dataDir as NSString).appendingPathComponent(configFileName)
            : opts.configPath
        let credsPath = opts.credentialsPath.isEmpty
            ? (dataDir as NSString).appendingPathComponent(credentialsFileName)
            : opts.credentialsPath

        if let token = tokenFromConfig(path: configPath) {
            return (token, .config)
        }
        if let token = tokenFromCredentials(path: credsPath) {
            return (token, .credentials)
        }
        return ("", .none)
    }

    /// Formats Authorization header value: `Bearer <token>` or empty when token is empty.
    static func authorizationHeader(token: String) -> String {
        token.isEmpty ? "" : "Bearer \(token)"
    }

    static func normalizeServer(_ server: String) -> String {
        var s = server.trimmingCharacters(in: .whitespacesAndNewlines)
        while s.hasSuffix("/") {
            s.removeLast()
        }
        return s
    }

    private static func tokenFromConfig(path: String) -> String? {
        guard FileManager.default.fileExists(atPath: path),
              let data = try? Data(contentsOf: URL(fileURLWithPath: path)),
              let cfg = try? JSONDecoder().decode(ConfigFile.self, from: data)
        else {
            return nil
        }

        // Prefer local loopback domain with non-empty trimmed token.
        for target in localLoopbackServers {
            if let t = domainTokenMatching(cfg.domains, wantNormalized: target) {
                return t
            }
        }

        // Else default domain.
        let def = normalizeServer(cfg.defaultServer)
        if !def.isEmpty, let t = domainTokenMatching(cfg.domains, wantNormalized: def) {
            return t
        }
        return nil
    }

    private static func domainTokenMatching(_ domains: [Domain], wantNormalized: String) -> String? {
        for d in domains {
            guard normalizeServer(d.server) == wantNormalized else { continue }
            let t = d.token.trimmingCharacters(in: .whitespacesAndNewlines)
            if !t.isEmpty {
                return t
            }
        }
        return nil
    }

    private static func tokenFromCredentials(path: String) -> String? {
        guard FileManager.default.fileExists(atPath: path),
              let content = try? String(contentsOfFile: path, encoding: .utf8)
        else {
            return nil
        }
        for line in content.split(separator: "\n", omittingEmptySubsequences: false) {
            let t = line.trimmingCharacters(in: .whitespacesAndNewlines)
            if !t.isEmpty {
                return t
            }
        }
        return nil
    }
}
